Optimize spherical manifold get_new_point into get_new_points

gassmoeller opened the pull request:
This PR is on top of #5529 and related to #5506.

First of all sorry for not coming back to this earlier. @masterleinad I checked #5529, and it indeed improves timings for large computations (when less points are computed in the Newton iteration), but it is still significantly slower than before. I suppose that is because now more instructions are done for each point, even for the simple case (e.g. the check for max_distance). I adapted the idea of @kronbichler in #5506 and moved all operations from `get_new_point` that do not depend on the weights into a new `SphericalManifold::get_new_points` functions. Although it is a pain to get all special cases correct, now I am essentially back to the performance before the introduction of the Newton iteration (while still having the increased accuracy of it). I show some timings below.  This new `get_new_points` functions could also allow further optimization as mentioned here https://github.com/dealii/dealii/issues/5506#issuecomment-345641636

Although it is hard to see, the logic in this PR should stay unchanged compared to #5529, except I now use the maximum distance between the surrounding points as upper bound for the maximum distance between candidate point and surrounding points. This seems reasonable as the candidate point lies in between the surrounding points. With this approximation the maximum distance check is independent of the weights.

Some timings below. All were taken from ASPECT, showing all steps of the assembly of an advection-diffusion system, for a 3D spherical shell (Rinner ~ 0.55 Router) with 4 refinements (393,216 cells, 3,195,010 Dofs for a Q2 element), and a MappingQGeneric(4) on 40 MPI processes.

Version | Time
------------ | -------------
8.5 |  6.3s
9.0 pre #5368   | 6.27s
9.0 pre #5529 | 9.27s
9.0 with #5529 | 7.24s
9.0 with this PR | 6.19s

I guess it is hard to generalize to all refinements and special cases, but for the moment I would be happy with this result. @masterleinad: I would appreciate if you could take a look at the code if I broke something (spherical manifold tests seem to work on my machine). Btw: I did not change anything in the Newton iteration itself, so if the discussion in #5506 leads to improvements in the Newton loop that might speed things up further.

kronbichler commented:
@gassmoeller This looks really nice at first sight. Thanks for following up on my idea, it is indeed what I had in mind. I will have a closer look with code-specific comments during the following days.

masterleinad commented:
/run-tests

masterleinad commented:
Thanks for doing this!
This reduces the cost for my modified step-10 testcase from `148*10^9` instructions (#5529) to `135*10^9` instructions. So even for rather fine meshes there still is like a 10% gain.

I'll reevaluate if the threshold 10^-2 is still good or can get a bit larger.

Apart from that we need to think on how to proceed with this. In the end, this just supersedes #5529.

masterleinad commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -235,6 +235,26 @@ class SphericalManifold : public Manifold<dim, spacedim>
+  void
+  get_new_points (const ArrayView<const Point<spacedim>> &surrounding_points,
+                  const Table<2,double>                  &weights,
+                  ArrayView<Point<spacedim>>              new_points) const;
```
Can you mark this with `override`?

masterleinad commented:
For me the following tests fail (locally):
```
     2085 - bits/fe_tools_02.debug (Failed)
     2087 - bits/fe_tools_03.debug (Failed)
     2559 - codim_one/mapping_fe_field_01.debug (Failed)
     2560 - codim_one/mapping_fe_field_01.release (Failed)
     4464 - grid/project_to_object_01.debug (Failed)
     4465 - grid/project_to_object_01.release (Failed)
     4486 - grid/torus_01.debug (Failed)
     4487 - grid/torus_01.release (Failed)
     5606 - manifold/chart_manifold_10.debug (Failed)
     5607 - manifold/chart_manifold_10.release (Failed)
     5848 - matrix_free/copy_feevaluation.debug (Failed)
     6014 - matrix_free/matrix_vector_stokes_noflux.debug (Failed)
     6016 - matrix_free/matrix_vector_stokes_notempl.debug (Failed)
```

In particular, the failing `matrix_free` tests report
```
--------------------------------------------------------
 6016: An error occurred in line <2894> of file <../source/fe/mapping_q_generic.cc> in function
 6016:     virtual CellSimilarity::Similarity dealii::MappingQGeneric<3, 3>::fill_fe_values(const typename Triangulation<dim, spacedim>::cell_iterator &, const CellSimilarity::Similarity,      const Quadrature<dim> &, const typename Mapping<dim, spacedim>::InternalDataBase &, internal::FEValues::MappingRelatedData<dim, spacedim> &) const [dim = 3, spacedim = 3]
 6016: The violated condition was:
 6016:     det > 1e-12*Utilities::fixed_power<dim>(cell->diameter()/ std::sqrt(double(dim)))
 6016: Additional information:
 6016:     The image of the mapping applied to cell with center [-0.364232 -0.364232 0.179005] is distorted. The cell geometry or the mapping are invalid, giving a non-positive volume          fraction of -0.00664757 in quadrature point 13.
 6016: --------------------------------------------------------
```

kronbichler commented:
@masterleinad I have not looked in detail, but there is a fix in d50e3419 in #5386 that will in turn also need #5564, so I don't think we should bother about those three tests here. (Unless it means that something fundamental in the spherical manifold is broken, of course.)

masterleinad commented:
@kronbichler Yes, applying https://github.com/dealii/dealii/commit/d50e34194b478019cd69afadcd54ea772e9f19c5 fixes these failing tests. Would you mind to cherry-pick this into a separate PR?

kronbichler commented:
Done.

masterleinad commented:
Thanks!

kronbichler commented on source/grid/manifold_lib.cc:
```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
+
+  // Step 1: Check for some special cases, create simple linear guesses otherwise.
+  const double tolerance = 1e-10;
+  std::vector<bool> accurate_point_was_found(new_points.size(),false);
```
Can you make this a `boost::container::small_vector`?

kronbichler commented on source/grid/manifold_lib.cc:
```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
-  const unsigned int n_points = vertices.size();
+  // To avoid duplicating all of the logic in get_new_points, simply call it
+  // for one position.
+  const Table<2,double> table_weights(1,weights.size(),weights.begin());
```
It is a bit unfortunate that we need to allocate memory here. Maybe create an internal function that does not take a `Table` as the argument but a plain pointer/`Arrayview`. (We will need to index into those in a somewhat more clumsy way, but @drwells worked hard to avoid memory allocations from these spots...)

kronbichler commented on source/grid/manifold_lib.cc:
```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
+
+  // Search for duplicate directions and merge them to minimize the cost of
+  // the get_new_point function call below.
+  Table<2,double> merged_weights(weights);
```
It would be great if we could avoid this memory allocation. We could use a small vector (of size say 400) here as well and manually index into the array by `size1*row+i` and the like...

kronbichler commented on source/grid/manifold_lib.cc:
```diff
@@ -446,33 +597,18 @@ get_new_point (const ArrayView<const Point<spacedim>> &vertices,
-            directions.push_back(direction_3d);
-            merged_weights.push_back(weights[i]);
-          }
+          directions_3d[i][c] = directions[i][c] / distances[i];
```
Do we still need to copy the data into a new array? Can we just use the original one? And can't we have `directions[j]` already normalization, i.e., on the unit sphere? (I might just have gotten lost by all the normalization steps we already do in this class.)

masterleinad replied:
The comment above seems to be outdated. The weights are already merged at this point.

masterleinad replied:
Regarding @kronbichler's comment: We already normalize in line 389, so we can turn this normalization (or the one in line 586) in an `Assert`.

masterleinad commented on source/grid/manifold_lib.cc:
```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
+  // Search for duplicate directions and merge them to minimize the cost of
+  // the get_new_point function call below.
+  Table<2,double> merged_weights(weights);
+  boost::container::small_vector<Tensor<1, spacedim>, 100> merged_directions(surrounding_points.size(),Point<spacedim>());
```
This should probably be `Tensor<1,spacedim>` instead of `Point<spcedim>`.

masterleinad commented on source/grid/manifold_lib.cc:
```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
+                const ArrayView<const double> &distances,
+                const ArrayView<const double> &weights) const
+{
+  const double tolerance = 1e-10;
```
We are using this value at several occasions in this class. I guess it makes sense to just turn it into a class variable.

masterleinad commented on source/grid/manifold_lib.cc:
```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
-  //
   // In this step, we consider all points and directions to be embedded
-  // in a three-dimensional space.
+  // in a three-dimensional space. The case spacedim < 2 was handled in get_new_points()
```
So we can specialize for `spacedim==3` and have an `Assert(false, ExcNotImplemented())` for in the general template. This way we can get rid off these annyoing copies.

masterleinad commented:
> I have one larger request/idea: After merging the weights "horizonally" (between different points), shouldn't we also be able to merge weights "vertically"? This would reduce the size of the loop in line 487 that does the get_new_point with the Newton iteration in the current version, but we would need to only return the unit point and add the different radii outside of that loop.

We should try that.

masterleinad commented:
With the new check for the maximum distance between `directions` the minimum distance that still requires a second Newton iterate is consistently around 0.113 from the second refinement on in my testcase.
The minimum distance for calls converging in the first Newton iterates are:
```
ref 1: 0.222
ref 2: 0.0507
ref 3: 0.0116
ref 4: 0.0273
```
and behaves as `ref^-2` resp. `h^2`. Hence, I would suggest to increase this tolerance a bit to 2.e-2 so that we benefit already in the third refinement. Since we are a factor 5 away from the minimum value in the second iterate, this should still be fine.

gassmoeller commented:
I think I addressed most comments except for the `tolerance` class member (because it is now used less often in the class, but instead in a non-member helper function that I need to do partial specialization).
The horizontal merge indeed significantly improves the speed of the Newton iteration (or rather calls the Newton iteration 5 times less often for a hyper shell with 0 refinement and MappingQ4), while the speed for the linear guess remains more or less constant. I suppose this is the best we can do for now. Please take a look if you approve of my specialization / function-call scheme.

masterleinad commented:
Hmm... I don't think that we can ignore the failing tests (by updating the output) at least for `mapping_q_mixed_manifolds_02` the diff
```
  DEAL::Volume 2D mapping degree 1: 0.156929 error: 0.00501399
 5680:   DEAL::Volume 2D mapping degree 2: 0.156152 error: 3.91474e-05
 5680: ! DEAL::Volume 2D mapping degree 3: 0.156146 error: 1.53930e-07
 5680: ! DEAL::Volume 2D mapping degree 4: 0.156146 error: 3.54084e-10
 5680: ! DEAL::Volume 2D mapping degree 5: 0.156146 error: 5.29174e-13
 5680: ! DEAL::Volume 2D mapping degree 6: 0.156146 error: -7.11016e-16
 5680:   DEAL::Volume 3D mapping degree 1: 0.0643409 error: 0.00501399
 5680:   DEAL::Volume 3D mapping degree 2: 0.0640224 error: 3.91474e-05
 5680:   DEAL::Volume 3D mapping degree 3: 0.0640199 error: 1.53930e-07
 5680: --- 1,10 ----
 5680:
 5680:   DEAL::Volume 2D mapping degree 1: 0.156929 error: 0.00501399
 5680:   DEAL::Volume 2D mapping degree 2: 0.156152 error: 3.91474e-05
 5680: ! DEAL::Volume 2D mapping degree 3: 0.156144 error: -1.09871e-05
 5680: ! DEAL::Volume 2D mapping degree 4: 0.156146 error: -1.10148e-07
 5680: ! DEAL::Volume 2D mapping degree 5: 0.156146 error: -1.66407e-08
 5680: ! DEAL::Volume 2D mapping degree 6: 0.156146 error: -1.69919e-10
 5680:   DEAL::Volume 3D mapping degree 1: 0.0643409 error: 0.00501399
 5680:   DEAL::Volume 3D mapping degree 2: 0.0640224 error: 3.91474e-05
 5680:   DEAL::Volume 3D mapping degree 3: 0.0640199 error: 1.53930e-07
```
shows a serious deterioration. Varying the tolerance for `max_distance` does not help here.

masterleinad commented:
Removing the last commit ("Merge horizontally") helps.

kronbichler commented:
@masterleinad The failure is in 2D, not in 3D, where we should not use `SphericalManifold`. Apart from the horizontal merging (which we still must verify carefully), I think we should fix the 2D issue on a more fundamental level by creating a specialization of `SphericalManifold` for `spacedim=2` that derives from `PolarManifold` (rather than using a member). That should be much cleaner.

masterleinad commented:
@gassmoeller Would you mind to rebase? Since #5564 is merged the number of failing tests should be much smaller.

masterleinad commented:
This patch at least fixes the 2D problems
```
diff --git a/source/grid/manifold_lib.cc b/source/grid/manifold_lib.cc
index 3bd2c3c..02afc31 100644
--- a/source/grid/manifold_lib.cc
+++ b/source/grid/manifold_lib.cc
@@ -474,14 +474,14 @@ get_new_points (const ArrayView<const Point<spacedim>> &surrounding_points,
       // If not in 3D, just use the implementation from PolarManifold
       // after we verified that the candidate is not the center.
       if (spacedim<3)
-        {
           new_points[row] = polar_manifold.get_new_point(surrounding_points, ArrayView<const double>(&weights[row*weight_columns],weight_columns));
-          accurate_point_was_found[row] = true;
-          continue;
-        }
-
     }
 
+  // In this case, we treated the case that the candidate is the center and obtained
+  // the new locations from the PolarManifold object otherwise.
+  if (spacedim<3)
+    return;
+
   // If all the points are close to each other, we expect the estimate to
   // be good enough. This tolerance was chosen such that the first iteration
   // for a at least three time refined HyperShell mesh with radii .5 and 1.
```

masterleinad commented:
@gassmoeller Can you please rebase? After #5386, there might be some changes in the test output. I would really like to get this merged soon.

gassmoeller commented:
Done. Sorry for the delay, I am currently at a conference with little time. Let's see if I merged the project_to_object results correctly.

masterleinad commented:
> Done. Sorry for the delay, I am currently at a conference with little time. 

Not a problem. I didn't mean to rush you.

> Let's see if I merged the project_to_object results correctly.

Well, it is still failing and `codim_one/mapping_fe_field_01.debug` is the other one.
`project_to_object` certainly looks reasonable and I am willing to believe that the other one is fine as well. 
Since I contributed some of the changes, I would like someone else to also approve this.

davydden commented:
```
The following tests FAILED:
	1780 - grid/project_to_object_01.debug (Failed)
Errors while running CTest
```

kronbichler commented:
Let us merge this anyway and see how the testers react. We can fix that one test afterwards if necessary.

gassmoeller commented:
Thanks for merging. 
I did not figure out how to use the reference tester results to update the test (my machine passed the test with differing results). @tjhei: Is there an easy way to do this? In aspect the tester provides a diff file that can be applied to the files locally, which is very convenient.

masterleinad commented:
So the test fails for you after #5672? Since [CDash](https://cdash.kyomu.43-1.org/testSummary.php?project=1&name=grid%2Fproject_to_object_01.debug&date=2018-01-01) doesn't report failure on any of the machines and setups, it would be interesting to hear about the setup you are using.
There is alternative outputfile that is labeled as `fused multiply-add`. I would not be surprised if this file needs to be updated as well. Do you think your setup would apply for this?

gassmoeller commented:
I just checked again, it passes after #5672. Not sure what happened there, I am sure I tested I several times before. Anyway, problem solved. Thanks!

tjhei commented:
> Is there an easy way to do this?

No, there is currently no easy way to do this except running the docker instance on your own machine (which I haven't documented either). Sorry.
